### PR TITLE
fixed bug when trying to read a task file and no permission

### PR DESCRIPTION
**Pull request by kforner**
Hello,

I had a bug when trying to backup files readable by root only, took me a little time to figure it out, so I wrote a little fix.
Moreover, none of your system calls are checked, which is evil ;-)
You could use autodie in the meantime.

Thanks for your work

### CODE REVIEW

**Comment by vsespb** [comment]
Hi. Case when file is not readable actually catched with exception on this line:
https://github.com/vsespb/mt-aws-glacier/blob/02c1ca3444223a9b22892549ed396a3c883b4cbc/lib/App/MtAws/FileUploadJob.pm#L56

But yo're right, this should be checked earlier for each file open and sane error message should be thrown.

I realize now that I was thinking about future code to handle complex cases (for example there can be race conditions between moment when file listing read and file
actually uploaded: if file was removed and directory with same name created - "open .. or confess" will not work at all. Or if file was just removed - we should not
die in this case - need to continue uploading other files), but forget to secure with exception simple cases like no read permissions.

I will merge it today. I will need to add other checks probably and bump version number imediately after fix.

Thanks for your report!

**Comment by kforner** [comment]
Hi

On Tue, Feb 19, 2013 at 9:36 AM, Victor Efimov notifications@github.comwrote:

> Hi. Case when file is not readable actually catched with exception on this
> line:
> 
> https://github.com/vsespb/mt-aws-glacier/blob/02c1ca3444223a9b22892549ed396a3c883b4cbc/lib/App/MtAws/FileUploadJob.pm#L56
> 
> But yo're right, this should be checked earlier for each file open and
> sane error message should be thrown.
> 
> I realize now that I was thinking about future code to handle complex
> cases (for example there can be race conditions between moment when file
> listing read and file
> actually uploaded: if file was removed and directory with same name
> created - "open .. or confess" will not work at all. Or if file was just
> removed - we should not
> die in this case - need to continue uploading other files), but forget to
> secure with exception simple cases like no read permissions.

Ok I understand. Maybe it does not have to be checked earlier, but in any
case as you said, you should have a clear message, e.g. warning "Skipping
file xxx that can not be read because : $!"

> I will merge it today. I will need to add other checks probably and bump
> version number imediately after fix.
> 
> Thanks for your report!

You're welcome !

**Comment by kforner** [comment]
Hi Victor,

I would need quickly some additional features from mtglacier, and I can
implement them or help you in implementing them.
What I would need is
- to be able to retrieve a list of files from a vault
- and one useful option would be a utility to display/select some files
  from the journal.log, with human readable dates and filename
- to be able to specify the options like --config anywhere in the
  command line, not just after the command

I presume that it is just a matter of implementing the interface, because
the actual processing is already there, and used by the _restore_ command.

It could be an additional --list=file|stdin option for restore.

What do you think ?

Best,

Karl

**Comment by vsespb** [comment]
> to be able to retrieve a list of files from a vault

you can do it now with retrieve-inventory/download-inventory commands. see README. it will create journal from vault content

> and one useful option would be a utility to display/select some files from the journal.log, with human readable dates and filename

filenames are readable. dates are not. you can use grep as journal is plain text file.

So, what exactly you want to implement?:
-dump journal as human readable data
-dump journal (only CREATED entries)
-dump journal but exclude CREATED enties if corresponding DELETED record exist
-implement some grep functionality when dumping journal

> to be able to specify the options like --config anywhere in the command line, not just after the command

Actual problem is: command (like sync/restore) should be specified as first argument. So you can use --config anywhere, except first argument.
This can be (easy?) fixed by dirty hack. Normal, good fix is not trivial.

Also I completelly rewroked command line options processing engine (as actually the future options/cli design that I am going to implement is pretty complex).
That problem is fixed in this rework. I hope merge it by end of Februarry.
So if you will be fixing something in options processing, don't submit pull request.

**Comment by kforner** [comment]
Thanks for replying quickly

On Tue, Feb 19, 2013 at 4:18 PM, Victor Efimov notifications@github.comwrote:

> to be able to retrieve a list of files from a vault
> 
> you can do it now with retrieve-inventory/download-inventory commands. see
> README. it will create journal from vault content

Sorry I was not clear, I meant: retrieve from the vault a given list of
files (for instance extracted from the inventory).
My use case is pretty standard I guess: I store all my backups in a vault.
If I want to restore from a given date, I have first to pull the
corresponding files from the vault.
I cannot afford to restore all missing files, that could be terabytes of
data.

>  and one useful option would be a utility to display/select some files
> from the journal.log, with human readable dates and filename
> 
> filenames are readable. dates are not. you can use grep as journal is
> plain text file.
> 
> So, what exactly you want to implement?:
> -dump journal as human readable data
> -dump journal (only CREATED entries)
> -dump journal but exclude CREATED enties if corresponding DELETED record
> exist
> -implement some grep functionality when dumping journal
> 
> Something like "-dump journal but exclude CREATED enties if corresponding
> DELETED record exist". I guess only two columns, tab separated: filename
> and date.
> Unless you know some command-line utility that can easily convert EPOCH
> seconds to dates.
> This way I could extract from it the list of files that I need to feed to
> the restore --list command I was talking about.
> Speaking of which, a dump --missing could be useful, listing only files
> that are not found locally.
> 
>  to be able to specify the options like --config anywhere in the command
> line, not just after the command
> 
> Actual problem is: command (like sync/restore) should be specified as
> first argument. So you can use --config anywhere, except first argument.
> This can be (easy?) fixed by dirty hack. Normal, good fix is not trivial.

I kinda remember that GetOpt::Long should be able to automatically extract
the options from anywhere in the command  line, leaving you with only the
positional arguments.
And to generate the documentation.

> Also I completelly rewroked command line options processing engine (as
> actually the future options/cli design that I am going to implement is
> pretty complex).
> That problem is fixed in this rework. I hope merge it by end of Februarry.
> So if you will be fixing something in options processing, don't submit
> pull request.

ok. anyway this is really not important.

Thanks again for this software that is really useful to me.

**Comment by vsespb** [comment]
> Sorry I was not clear, I meant: retrieve from the vault a given list of files (for instance extracted from the inventory).

ok. get it.

> My use case is pretty standard I guess: I store all my backups in a vault. If I want to restore from a given date,
> I have first to pull the corresponding files from the vault. I cannot afford to restore all missing files, that could be terabytes of data.

You can accompilish this by dirty hack right now - use grep on journal. Create another journal file only with files that you need. And restore.
Or you can sort files in journal by importance and restore with --max-number-of files option.

But if you want to implement this in code and submit pull request, please use option names compatible with following behaviour:

one of more options like 
--include "pattern"
--exclude "pattern"
order is important.
default is include all. if you wan't exlucde to be default, need to add --exlucde "**" as last option. (I am not sure about this point, maybe need to provide
another option like --exclude-by-default. Or maybe use exclude is default behaviour for restore command and include-is-default for sync command. not sure).

also not sure about wildcards.

also use --include-from --exclude-from - same, but takes list from file

behaviour similar to rsync and duplicity http://man.cx/rsync%281%29#heading15

and please submit request to "development" branch, not master. master is only for releases and emergency fixes.

> Something like "-dump journal but exclude CREATED enties if corresponding DELETED record exist"

Not implemented yet. If you implement it and submit pull request, please name it "list-files --journal myjournal". Also please submit to "development" branch.

> Unless you know some command-line utility that can easily convert EPOCH seconds to dates.

perl -e 'print time'|perl -MTime::Piece -ne 'print join(";",localtime($_)->datetime)."\n"'
(output is date in local timezone)

> Speaking of which, a dump --missing 

That's better to implement in check-local-hash command (to be renamed btw). check-local-hash --format=print --missing

> I kinda remember that GetOpt::Long should be able to automatically extract the options from anywhere in the command line

Right. But GetOpt::Long is only works when you supply full list of possible options. So I have now separate list of options for each command. That is design flaw.
Thus I have to get command name before actually parsing options. New engine based on GetOpt::Long too. This problem is solved there.
Basically I found that sane command line interface is one of most hard part of this program. So I am creating new engine (wrapper) over GetOpts to deal with validation of options
and group of options which are not compatible in some combinations, which can be used with different commands, and when mandatory/optional status of option depends on other options.

**Comment by vsespb** [comment]
UPD:

> Or you can sort files in journal by importance and restore with --max-number-of files option.

actually no, you can't. so you can only use grep to extract needed files from one journal to another.

**Comment by kforner** [comment]
OK. I will not implement anything before the new "options" implementation
is done, to avoid useless merge problems.
I can live with your journal hacking work-around in the meantime.

But I managed to convince myself that a multi-script solution could be
nice, this way we can safely write a mtglacier_journal script to
parse/select/print the journal
without any impact on the main script.

Best,
Karl

**Comment by vsespb** [comment]
> Don't you think an easy way would simply be to split the script into several ones,

I think it won't help. Even if people split script into several executable commands, it's often symlinks to one executable, which parses command line and detects
what command was ran.

Also lots of options will be shared between different commands. 

ps.

I am not Git guru, but I feel like Git is not example of clean command line option design, also not an example of good command line parsing implementation.

recently I coded script
https://github.com/vsespb/git-incremental-backup/blob/a770918bfad4ce779764935d72f761e1016954d1/git-inc-backup.pl#L64

and found a known bug in Git
http://stackoverflow.com/questions/11292057/git-windows-git-pull-cannot-be-used-without-a-working-tree

that's exactly the case when same option shared between several commands and sometimes it does not work..

> But I managed to convince myself that a multi-script solution could be nice, this way we can safely write a mtglacier_journal script to parse/select/print the journal

I have Journal.pm which reads journal and writes journal (including old versions). So mtglacier_journal will still have to use it.. otherwise it's not DRY (same code repeated twice). Also I have unit tests that ensure that journal parsed right with this code. Finally FreeBSD users were complaining that they use non-Unicode encoding on their systems, so they want to have support for Journal in different encodings (to work with grep etc too). So journal processing will have to handle encoding option/config variable.

Well, actually it's hard to explain what problems I had with command line, as future command/options to be implemented are only in my head and some draft notes, but not documentation, and need to show full picture to explain this problem..

> OK. I will not implement anything before the new "options" implementation is done, to avoid useless merge problems.

Yep there can be. Yes if hacks with grep work fine for you, it's better just to use hacks for now.

p.s.
Also, forgot to say, just in case, if you going to submit pull request, don't use CPAN modules. Only core modules. Should be compatible with Perl 5.8.8
(and modules should be in core since 5.8.8). CentOS 5.x (which is supported till 2017) still uses Perl 5.8.8.
And one if goal is simple deploy, with OS package manager, without CPAN.
I didn't have time to write guide for contributors and put this into it..

**Comment by kforner** [comment]
Hi Victor,

In my multi-script "vision", each script should have minimal code in it.
I once wrote a perl toolkit for dealing with tabular data files such as
excel, csv, tsv.
The interface of the toolkit was composed of many scripts, such as:
sheet_read, shead_header, sheet_join, sheet_print, sheet_info, sheet_cut,
sheet_select and so on.

Here's the _sheet_cut.pl_ script as an example:
As you can see the script is pretty minimal.
I have a module that deals with common options, and that can process
additional options (_processStandardOptions( options => [qw(right inner)] );
*There is no code duplication. I could have add the few lines of processing
code in a package but it was quite trivial._
*
#! /usr/bin/env perl
use FindBin qw($RealBin);
use lib "$RealBin/../../lib";
use Filter::Include;
include Me::Sheet::HandyModules;

_my %OPTIONS = processStandardOptions( options => [qw(right inner)] );_

my $left_sheet = getAndLoadDefaultSheet; # modify @ARGV

my ( $right_file, $left_col, $right_col, $left_columns_to_keep,
$right_columns_to_keep )
    = @ARGV;
pod2usage("give right_file left_col right_col") unless defined $right_col;

my $right_sheet = Me::Sheet->MAGIC_READ_FILE($right_file);

my %opts;
$opts{left_columns_to_keep} = [processColumnsOption( $left_sheet,
$left_columns_to_keep )] if defined $left_columns_to_keep;
$opts{right_columns_to_keep} = [processColumnsOption( $right_sheet,
$right_columns_to_keep)] if defined $right_columns_to_keep;
$opts{inner} = $OPTIONS{inner};

my $joined_sheet;
if ( $OPTIONS{right} ) {
    $joined_sheet = $left_sheet->right_join(
        $right_sheet, $left_col, $right_col,\%opts);
} else {
    $joined_sheet = $left_sheet->left_join(
        $right_sheet, $left_col, $right_col,\%opts);
}
$joined_sheet->magic_print;

**Comment by vsespb** [comment]
Well, yes, I can code like this. problem is that I will have more options and more complicated conditions and validations.
so I wrote DSL (new engine), example here https://github.com/vsespb/mt-aws-glacier/blob/getopt_rework/lib/App/MtAws/ConfigDefinition.pm implementation here https://github.com/vsespb/mt-aws-glacier/blob/getopt_rework/lib/App/MtAws/ConfigEngineNew.pm (just pushed, not finished)
that's just a small part of options and validations that I am going to implement.

Also I deal with things like unit tests for each released command line API (API always should be backward compatible) which a) should not break if I just change error message text, b) should be easy to write (so preprocessing options is preferred on raising exceptions in the middle of execution)

Examples https://github.com/vsespb/mt-aws-glacier/blob/getopt_rework/t/integration/config_engine_v078.t https://github.com/vsespb/mt-aws-glacier/blob/getopt_rework/t/integration/config_engine_v082.t

Other things which add complication are: Error messages, Warning messages, Deprecated options and commands. Renamed options and commands, Options in config. Options which require additional libraries. Default values for options, Option values with non ASCII characters. Positional arguments. Going
to support environment variables and several config files.

Spliting code to several programs and reuse code actually a good idea (code that not used in current command is not loaded), but this can be accompilished also by perl 'require'.
Also this breaks compatibility and makes deploy more complicated.
